src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala

```diff
@@ -358,29 +434,37 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
 // mentioned in the body. This is useful to translate to SQL (join
 // conditions, select, etc.)
 class QuerySchema(q : ConjunctiveQuery) {
-    var query_schema = new HashMap[ String, Tuple2[Int,Variable] ]()
+  var query_schema = new HashMap[ String, Tuple2[String,Variable] ]()
 
   // maps each variable name to a canonical version of itself (first occurence in body in left-to-right order)
   // index is the index of the subgoal/atom this variable is found in the body.
   // variable is the complete Variable type for the found variable.
   def generateCanonicalVar()  = {
-    q.bodies(0).zipWithIndex.foreach {
-      case (Atom(relName,terms),index) =>  {
-        terms.zipWithIndex.foreach { case (expr, i) =>
-          expr match {
-            case VarExpr(v) => 
-              if (! (query_schema contains v) )
-                query_schema += { v -> (index, Variable(v,relName,i) ) }
-            case _ =>
-          }
+    def generateCanonicalVarFromAtom(a: Atom, index: String) {
+      a.terms.zipWithIndex.foreach { case (expr, i) =>
+        expr match {
+          case VarExpr(v) =>
+            if (! (query_schema contains v) )
+              query_schema += { v -> (index, Variable(v,a.name,i) ) }
+          case _ =>
         }
       }
     }
+
+    def generateCanonicalVarFromBodies(bodies: List[Body], indexPrefix: String) {
```

Comment:
Could you comment how this is handling ModifierAtom specially, and why that's necessary?